### PR TITLE
Use SOS to dump managed stack traces from a dump on Windows

### PR DESCRIPTION
**Pull request by jkoritzinsky**
Download SOS as part of our Helix jobs and load it into cdb to dump managed stack traces when a test crashes or times out.

Also introduces a synthetic test failures to validate behavior (will be removed before merging).

### CODE REVIEW

File: src/tests/Common/helixpublishwitharcade.proj [hunk 1]
```diff
@@ -723,6 +723,10 @@
   <ItemGroup Condition=" '$(UsesHelixSdk)' == 'true' ">
     <HelixCorrelationPayload Include="$(CoreRootDirectory)" />
     <HelixCorrelationPayload Include="$(XUnitLogCheckerDirectory)" />
+    <HelixCorrelationPayload Condition="'$(TestWrapperTargetsWindows)' == 'true'" Include="dotnet-sos">
+      <Destination>sos</Destination>
+      <Uri>https://pkgs.dev.azure.com/dnceng/public/_packaging/dotnet-tools/nuget/v3/flat2/dotnet-sos/$(DotnetSosVersion)/dotnet-sos.$(DotnetSosVersion).nupkg</Uri>
```
Comment (stephentoub) [inline src/tests/Common/helixpublishwitharcade.proj:728]: Before dotnet-sos existed, we used clrmd directly in RemoteExecutor:
https://github.com/dotnet/arcade/blob/bdc59254cf108e1d48451dc43bb9ebc331cdca7b/src/Microsoft.DotNet.RemoteExecutor/src/RemoteInvokeHandle.cs#L177-L225
We might want to standardize on one or the other.

Comment (jkoritzinsky) [reply src/tests/Common/helixpublishwitharcade.proj:728]: ClrMD there is doing process attach only for timeouts, whereas we're handling both timeouts and crashes here, so it's a little different. Definitely worth considering consolidation though.

Comment (jkoritzinsky) [reply src/tests/Common/helixpublishwitharcade.proj:728]: There were some talks about standardizing this sort of mechanism (dumping stack traces from a dump to stdout for the build analysis tooling), but with the recent changes to the EngSrv teams, I don't know how much of that work will still happen.

Comment (stephentoub) [reply src/tests/Common/helixpublishwitharcade.proj:728]: Right, I understand they're handling different cases.  But they can both handle both, so it's overkill to have two different tools used for the same purpose.  Simply suggesting we choose one and stick with it.

Comment (jkoritzinsky) [reply src/tests/Common/helixpublishwitharcade.proj:728]: I'll look at the RemoteExecutor implementation and see if we can consolidate something useful.

Comment (stephentoub) [reply src/tests/Common/helixpublishwitharcade.proj:728]: Cool, thanks.

File: src/tests/Common/Coreclr.TestWrapper/CoreclrTestWrapperLib.cs [hunk 2]
```diff
@@ -601,10 +601,14 @@ static bool TryPrintStackTraceFromDmp(string dmpFile, StreamWriter outputWriter)
                 return false;
             }
 
+            string sosPath = Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.UserProfile), ".dotnet", "sos", "sos.dll");
+
             var cdbScriptPath = Path.GetTempFileName();
             // TODO: Add SOS support once we can easily download SOS to install.
-            File.WriteAllText(cdbScriptPath, """
+            File.WriteAllText(cdbScriptPath, $$"""
```
Comment (danmoseley) [inline src/tests/Common/Coreclr.TestWrapper/CoreclrTestWrapperLib.cs:608]: Remove todo above?

File: src/tests/Common/helixpublishwitharcade.proj (hunk 1 above)
Comment (jkoritzinsky) [reply src/tests/Common/helixpublishwitharcade.proj:728]: We'll need to use something like this pattern to get both native and managed stack traces, and we'll likely want to use this model as well to solve https://github.com/dotnet/runtime/issues/83047 (which focuses more on crashes than timeouts) if we don't move to using `dotnet test` for the libraries tests in Helix. The RemoteExecutor version has a much better UX around the output though as it has more structure to work with (from CLRMD's APIs).

**Comment by jkoritzinsky** [comment]
Marking this ready for review as the change works (see the logs for the CoreCLR windows runtime tests), but marked as no-merge as I need to remove the induced test failure.

@hoyosjs for review.

File: src/tests/Common/helixpublishwitharcade.proj [hunk 3]
```diff
@@ -865,4 +869,10 @@
   <Import Sdk="Microsoft.DotNet.Helix.Sdk" Project="Sdk.targets" Condition=" '$(UsesHelixSdk)' == 'true' " />
   <Import Sdk="Microsoft.Build.NoTargets" Project="Sdk.targets" Condition=" '$(UsesHelixSdk)' != 'true' " />
 
+  <!-- This target needs to come after importing the Helix SDK as AfterTargets doesn't work for targets that have yet to be defined. -->
+  <Target Name="ConfigureSOS" AfterTargets="AddDotNetSdk" BeforeTargets="CoreTest" Condition="'$(TestWrapperTargetsWindows)' == 'true' and '$(RuntimeFlavor)' == 'CoreCLR'">
+    <PropertyGroup>
+      <HelixPreCommands>$(HelixPreCommands);dotnet %25HELIX_CORRELATION_PAYLOAD%25\sos\tools\net$(DotnetSosTargetFrameworkVersion)\any\dotnet-sos.dll install</HelixPreCommands>
```
Comment (hoyosjs) [inline src/tests/Common/helixpublishwitharcade.proj:875]: We may want to explicitly specify the architecture just to be sure.

**Review by hoyosjs — APPROVED** [review]

File: src/tests/Common/helixpublishwitharcade.proj (hunk 3 above)
Comment (jkoritzinsky) [reply src/tests/Common/helixpublishwitharcade.proj:875]: Is the architecture parameter in `dotnet-sos` case-sensitive? Just want to know before I mess with it.

Comment (jkoritzinsky) [reply src/tests/Common/helixpublishwitharcade.proj:875]: Looks like it's not. I'll add the architecture.

**Comment by hoyosjs** [comment]
@jkoritzinsky this looks good now. It's not loading the PDBs for managed for for line numbers, but that's separate.

**Comment by jkoritzinsky** [comment]
Networking failures are known issues, wasm timeout is unrelated. Merging this in. I'll file an issue to follow-up on providing a consistent story for dumping stacks on crashes.

**Pull request merged**